docs/kb/wasm.rst

```diff
@@ -0,0 +1,307 @@
+
+UDF's source code can be, naturally, simply coded by hand in ``WAT``. It is not often very convenient to program directly in assembly, so here are a few tips.
+
+Compiling to Wasm
+^^^^^^^^^^^^^^^^^
+
+Rust
+....
+
+The main supported language for Wasm UDFs is Rust. To generate WebAssembly from rust, it's best to use the `scylla-udf <https://github.com/scylladb/scylla-rust-udf>`_ Rust helper library, and follow the instructions present there.
+
+As a short example, here's a sample Rust code which can be compiled to WebAssembly:
+
+.. code-block:: rust
+
+    use scylla_udf::export_udf;
+
+    #[export_udf]
+    pub fn fib(n: i32) -> i32 {
+        if n < 2 {
+            n
+        } else {
+            fib(n - 1) + fib(n - 2)
+        }
+    }
+
+The compilation instructions are described at https://github.com/scylladb/scylla-rust-udf but the commands will generally be:
+
+.. code-block:: bash
+
+    cargo build --target=wasm32-wasi
+    wasm2wat target/wasm32-wasi/debug/fib.wasm > fib.wat
+
+
+C
+.
+
+Clang is capable of compiling C source code to Wasm and it also supports useful built-ins
+for using Wasm-specific interfaces, like ``__builtin_wasm_memory_size`` and ``__builtin_wasm_memory_grow``
+for memory management.
+
+However, there is no C helper library yet, so implementing UDFs in C is in general much more difficult than in Rust.
+Just to implement the fib() function, you need something like this:
+
+.. code-block:: c
+
+    #include<stdlib.h>
+
+    const int WASM_PAGE_SIZE = 64 * 1024;
+
+    const int _scylla_abi = 2;
+
+    void* _scylla_malloc(int size) {
+        return malloc(size);
+    }
+
+    void _scylla_free(void* ptr) {
+        free(ptr);
+    }
+
+    static long long swap_int64(long long val) {
+        val = ((val << 8) & 0xFF00FF00FF00FF00ULL ) | ((val >> 8) & 0x00FF00FF00FF00FFULL );
+        val = ((val << 16) & 0xFFFF0000FFFF0000ULL ) | ((val >> 16) & 0x0000FFFF0000FFFFULL );
+        return (val << 32) | ((val >> 32) & 0xFFFFFFFFULL);
+    }
+
+    long long fib_aux(long long n) {
+        if (n < 2) {
+            return n;
+        }
+        return fib_aux(n-1) + fib_aux(n-2);
+    }
+
+    long long fib(long long p) {
+        int size = p >> 32;
+        long long* p_val = (long long*)(p & 0xffffffff);
+        // Initialize memory for the return value
+        long long* ret_val = _scylla_malloc(sizeof(long long));
+        if (size == -1) {
+            *ret_val = swap_int64(42);
+        } else {
+            *ret_val = swap_int64(fib_aux(swap_int64(*p_val)));
```

Comment:
This swap looks strange but it actually just swaps the byte order of the 64 bit int. We're doing this because we decided to use the serialization from Scylla for writing values into the Wasm memory, and we also swap the bye order in Scylla. Would it be better if I renamed the "swap" to for example `ntohll`?